File: examples/julia/web3-example/eth_client.jl
# Julia Ethereum Client
# High-performance scientific computing for Web3

using HTTP
using JSON

struct EthClient
    rpc_url::String
end

function rpc_call(client::EthClient, method::String, params::Vector=[])
    body = Dict(
        "jsonrpc" => "2.0",
        "method" => method,
        "params" => params,
        "id" => 1
    )

    response = HTTP.post(
        client.rpc_url,
        ["Content-Type" => "application/json"],
        JSON.json(body)
    )

    result = JSON.parse(String(response.body))
    return result["result"]
end

function get_block_number(client::EthClient)
    hex = rpc_call(client, "eth_blockNumber")
    return parse(Int64, hex, base=16)
end

function get_balance(client::EthClient, address::String)
    return rpc_call(client, "eth_getBalance", [address, "latest"])
end

function get_gas_price(client::EthClient)
    return rpc_call(client, "eth_gasPrice")
end

function wei_to_ether(wei_hex::String)
    wei = parse(BigInt, wei_hex, base=16)
    return Float64(wei) / 1e18
end

# Example usage
function main()
    client = EthClient("https://eth-mainnet.g.alchemy.com/v2/YOUR-API-KEY")
    vitalik = "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045"

    block_number = get_block_number(client)
    balance = get_balance(client, vitalik)
    balance_eth = wei_to_ether(balance)

    println("Block Number: $block_number")
    println("Balance: $balance_eth ETH")

    # Performance analysis
    @time for _ in 1:100
        get_block_number(client)
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: examples/fsharp/web3-example/EthClient.fs
// F# Ethereum Client
// Functional Web3 on .NET

module EthClient

open System
open System.Net.Http
open System.Text
open System.Text.Json

type RpcRequest = {
    jsonrpc: string
    method: string
    ``params``: obj[]
    id: int
}

type RpcResponse = {
    result: JsonElement
}

let httpClient = new HttpClient()

let rpcCall (rpcUrl: string) (method: string) (parameters: obj[]) =
    async {
        let request = {
            jsonrpc = "2.0"
            method = method
            ``params`` = parameters
            id = 1
        }

        let json = JsonSerializer.Serialize(request)
        let content = new StringContent(json, Encoding.UTF8, "application/json")

        let! response = httpClient.PostAsync(rpcUrl, content) |> Async.AwaitTask
        let! body = response.Content.ReadAsStringAsync() |> Async.AwaitTask

        let result = JsonSerializer.Deserialize<RpcResponse>(body)
        return result.result
    }

let getBlockNumber rpcUrl =
    async {
        let! result = rpcCall rpcUrl "eth_blockNumber" [||]
        let hex = result.GetString()
        return Convert.ToInt64(hex, 16)
    }

let getBalance rpcUrl address =
    async {
        let! result = rpcCall rpcUrl "eth_getBalance" [| address; "latest" |]
        return result.GetString()
    }

let getGasPrice rpcUrl =
    async {
        let! result = rpcCall rpcUrl "eth_gasPrice" [||]
        return result.GetString()
    }

[<EntryPoint>]
let main argv =
    let rpcUrl = "https://eth-mainnet.g.alchemy.com/v2/YOUR-API-KEY"
    let vitalik = "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045"

    async {
        let! blockNumber = getBlockNumber rpcUrl
        let! balance = getBalance rpcUrl vitalik
        let! gasPrice = getGasPrice rpcUrl

        printfn "Block Number: %d" blockNumber
        printfn "Balance: %s" balance
        printfn "Gas Price: %s" gasPrice
    }
    |> Async.RunSynchronously

    0
